pkg/operator/operator.go

```diff
@@ -598,10 +599,33 @@ func (op *Reporting) Run(ctx context.Context) error {
 	}
 	op.logger.Info("writes to Presto are succeeding")
 
+	// In the case where the Prometheus metrics importer configuration has not
+	// been disabled, try to fire off a test Prometheus query against the `up`
+	// metric. This should weed out any invalid Prometheus configurations, e.g.
+	// the service account token or CA passed to the reporting-operator is invalid,
+	// and we should complain about that, but not require this as part of the
+	// reporting-operator health checks. Note: in the case where we encounter
+	// something like a timeout error while polling, just log the message
+	// indicating there was a problem and ignore the error.
+	if !op.cfg.DisablePrometheusMetricsImporter {
+		op.logger.Info("testing the ability to query from Prometheus")
+		err = wait.Poll(3*time.Second, 15*time.Second, func() (bool, error) {
```

Comment:
A couple questions with this implementation.  I am understanding that the objective here is to test that the configuration we have for prometheus is correct.  With that assumption the questions are:

1.  Should we attempt to query prom multiple times - ie. will an incorrect configuration, network issue, etc resolve itself in the 15 seconds we're polling?
2.  Do we care about the actual state of prometheus at this point (the result of the query)?
3.  Do we care about the length of time the query to prometheus takes?

On the assumption that the answer to 2 is no (we're just logging the results of errors and prom going down at runtime needs to be accounted for anyway) and the answer to 3 is yes there are two suggestions.

First, no matter the answer to 1 - we cap the prom query by using a real context via a `context.TODO`.

If the answer to 1 is yes - bad config could resolve itself in the 15 second polling period we should adjust to not return an error (which stops polling right away) and log in the poll method, letting it time out if we don't resolve by the time `time.Duration` hits (15 seconds):

```go
err = wait.Poll(3*time.Second, 15*time.Second, func() (bool, error) {
   _, err := op.promConn.Query(context.TODO(), "up", time.Now())
   if err != nil {
      // give it another try, don't return the error
      op.logger.Errorf("failed to successfully query Prometheus: %v", err)
      return false, nil
   }

   return true, nil
})
```

If the answer to 1 is no, config will not change in the 15 seconds so we have no need to keep retrying it and we're just going to log unavailable prometheus anyway this boils down to just:

```go
_, err := op.promConn.Query(context.TODO(), "up", time.Now())
if err != nil {
   op.logger.Errorf("failed to successfully query Prometheus: %v", err)
}
```

Reply:
The main objective I had in mind for this code block is to provide some sort of logging or event that can provide an avenue for quickly debugging potential problems with our, or the user-provided, Prometheus configuration. With that in mind, firing off a query for the `up` metric in Prometheus at the beginning of the operator's workflow seemed like a decent health check. Note: we could uncover a subset of these errors while processing the ReportDataSource promQL queries too, so this kind of check could be redundant, but those failures could be due to something that's not configuration-based (e.g. a many-to-many error).

I wanted to account for a potential bad query to Prometheus (for whatever reason) so that's why I used `poll.Wait` to fire off subsequent ones, but the 15-second timeout was arbitrarily chosen with the goal that our operator would still pass the reporting-operator's health checks as we're not polling for a substantial amount of time. But to answer your questions, the configuration passed to the reporting-operator most likely isn't going to change within a 15-second window, so a single Prometheus query could suffice, assuming we update the context to have a deadline.

Reply:
And we could also fire off an event indicating a potential problem, on top of complaining about this failed query in the reporting-operator container logs.

Reply:
cool, thanks.  Ok, so if you want to try multiple times for whatever reason (maybe there is a network error, whatever) then not returning the error (like in the first example) will keep you polling like you wanted :+1: 

Other items of note.  You may consider changing `op.logger.Errorf` to warning level and also logging a success before `return true, nil` so when reading the log you can see if it ever resolved successfully (maybe even with the `result.String()` if that's helpful info).  

Thanks for the clarity.

Reply:
Made a couple of changes based on your feedback and tried to improve the logging. In the case where the host of the Prometheus instance is invalid, the following logs would be produced:
```
time="2020-07-30T15:39:09Z" level=info msg="testing the ability to query from Prometheus" app=metering
time="2020-07-30T15:39:16Z" level=warning msg="failed to succesfully query Prometheus: Get https://thanos-querier.openshift-monitoring.svc:8080/api/v1/query?query=up&time=2020-07-30T15%3A39%3A12.983041081Z: dial tcp 172.30.245.171:8080: connect: no route to host" app=metering
time="2020-07-30T15:39:19Z" level=warning msg="failed to succesfully query Prometheus: Get https://thanos-querier.openshift-monitoring.svc:8080/api/v1/query?query=up&time=2020-07-30T15%3A39%3A18.983069575Z: dial tcp 172.30.245.171:8080: connect: no route to host" app=metering
time="2020-07-30T15:39:25Z" level=warning msg="failed to succesfully query Prometheus: Get https://thanos-querier.openshift-monitoring.svc:8080/api/v1/query?query=up&time=2020-07-30T15%3A39%3A21.983024157Z: dial tcp 172.30.245.171:8080: connect: no route to host" app=metering
time="2020-07-30T15:39:28Z" level=warning msg="failed to succesfully query Prometheus: Get https://thanos-querier.openshift-monitoring.svc:8080/api/v1/query?query=up&time=2020-07-30T15%3A39%3A25.064029555Z: dial tcp 172.30.245.171:8080: connect: no route to host" app=metering
time="2020-07-30T15:39:28Z" level=warning msg="queries from Prometheus are failing: timed out waiting for the condition" app=metering
```
I'm not sure if every failure should be a warning level, but that's probably better than making them debug level, seeing that `queries from Prometheus are failing: timed out waiting for the condition` isn't super helpful and would require redeploying the reporting-operator with a higher log verbosity.

Reply:
Alternatively, could just return the error we grab from the poll.Wait condition function instead of nil, so the outer scope can just report on that error instead of the `timed out waiting for the condition`.

Reply:
> Alternatively, could just return the error we grab from the poll.Wait condition function instead of nil, so the outer scope can just report on that error instead of the timed out waiting for the condition.

If you return the error you stop polling so lose the "try multiple times" component depending on when the error happens.  You could get fancier and aggregate the errors (but not return them) and report it after poll finishes if it doesn't finish with success (where you nil out the aggregated error on success or something)

Reply:
but fwiw, it seems ok to be simple, to me.